compiler/rustc_mir_build/src/build/custom/parse/instruction.rs

```diff
@@ -139,6 +139,7 @@ impl<'tcx, 'body> ParseCtxt<'tcx, 'body> {
     }
 
     fn parse_rvalue(&self, expr_id: ExprId) -> PResult<Rvalue<'tcx>> {
+        #![allow(unreachable_patterns)]
```

Comment:
why are these needed when they weren't before?

Also please attach the attribute to the function instead of using an inner attribute

Reply:
Since I changed how the spans work (not entirely sure how/why), it will emit an error like this:
```rust
error: unreachable pattern
   --> compiler\rustc_mir_build\src\build\custom\parse.rs:54:13
    |
22  | / macro_rules! parse_by_kind {
23  | |     (
24  | |         $self:ident,
25  | |         $expr_id:expr,
...   |
54  | |             _ => return Err($self.expr_error(expr_id, $expected))
    | |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this arm is never executed
55  | |         }
56  | |     }};
57  | | }
    | |_- in this expansion of `parse_by_kind!`
    |
   ::: compiler\rustc_mir_build\src\build\custom\parse\instruction.rs:146:9
    |
146 | /         parse_by_kind!(self, expr_id, _, "rvalue",
147 | |             @call("mir_discriminant", args) => self.parse_place(args[0]).map(Rvalue::Discriminant),
148 | |             @call("mir_checked", args) => {
149 | |                 parse_by_kind!(self, args[0], _, "binary op",
...   |
171 | |             _ => self.parse_operand(expr_id).map(Rvalue::Use),
    | |             ^ this pattern is irrefutable; subsequent arms are never executed
172 | |         )
    | |_________- in this macro invocation
    |
    = note: `-D unreachable-patterns` implied by `-D warnings`
```

Reply:
please allow this lint inside the macro instead of at the call sites then.

Reply:
I just realized why this happens, it's because I now use the HirId of the scrutinee rather than the match arms, so I guess (the documentation doesn't say) this means the allow attributes now have to be on the match construct itself rather than individual arms.

Actually that seems really undesirable to me so unless there is a way around this that I didn't see, I'm not sure this PR is a good idea.

Reply:
Oh good catch. I guess we could use the hir id of the entire match instead?

Reply:
We need to run crater to see how this will impact the ecosystem.

Reply:
> We need to run crater to see how this will impact the ecosystem.

From some basic searches I can tell there will be many matches that will need to shift their `allow`s. Most of them are to deal with conditionally compiling match arms.

Reply:
Which I really don't like, actually. I think the allows should remain on the match arms.